In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [5]:
df_prods = pd.read_csv(r'C:\Users\kenta\Google ドライブ\CareerFoundry Immersion\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [20]:
df_ords = pd.read_csv(r'C:\Users\kenta\Google ドライブ\CareerFoundry Immersion\Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col = False)

In [7]:
# 1.If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

2. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

A. The Max cell in the Price column is extraordinary higher.

In [8]:
# 3. Check for mixed-type data in your df_ords dataframe. No mix type of data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

In [9]:
# 5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum() 

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [10]:
df_ords.shape

(3421083, 7)

In [11]:
pd.options.display.float_format = '{:.2f}'.format

In [12]:
df_ords.describe()

order_id    user_id  order_number  order_dow  order_hour_of_day  \
count 3421083.00 3421083.00    3421083.00 3421083.00         3421083.00   
mean  1710542.00  102978.21         17.15       2.78              13.45   
std    987581.74   59533.72         17.73       2.05               4.23   
min         1.00       1.00          1.00       0.00               0.00   
25%    855271.50   51394.00          5.00       1.00              10.00   
50%   1710542.00  102689.00         11.00       3.00              13.00   
75%   2565812.50  154385.00         23.00       5.00              16.00   
max   3421083.00  206209.00        100.00       6.00              23.00   

       days_since_prior_order  
count              3214874.00  
mean                    11.11  
std                      9.21  
min                      0.00  
25%                      4.00  
50%                      7.00  
75%                     15.00  
max                     30.00

In a markdown cell, report your findings and propose an explanation for any missing values you find.

206209 data of days_since_prior_order is missing out of 3421083, which is the relatively high proportion of the column. Average days_since_prior_order is 11.11 days, the median is seven days, the max is 30, and the minimum is 0 day.

In [13]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     nan  
1                   15.00  
2                   21.00  
3                   29.00  
4                   28.00

In [23]:
df_ords[df_ords['days_since_prior_order'].isnull()]

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     nan  
11                      11                     nan  
26                      14                     nan  
39                      11                     nan  
45                      12                     nan  
...                    ...                     ...  
3420930                 12                     nan  
3420934                 18                     nan  
3421002                 19                     nan  
3421019                 15                     nan  
3421069                 11                     nan  

[206209 rows x 7 columns]

In [24]:
df_ords['order_number'].value_counts()

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

# 6. Address the missing values using an appropriate method.

In a markdown cell, explain why you used your method of choice.

I just left it as it is because all data that shows nan is their 1st order, which has no prior deal before the purchase.

In [65]:
# 7. Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords[df_ords.duplicated()]

In [66]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No dupe

In [70]:
path = r'C:\Users\kenta\Google ドライブ\CareerFoundry Immersion\Instacart Basket Analysis'

In [72]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleansed.csv'))
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleansed.csv'))